### ISSUE TITLE
"memory regions" on XC

### ISSUE THREAD

**Issue by reuster986** [mhmerrill/arkouda#271]
@bradcray and @ronawho , occasionally when I run arkouda on a Cray XC, it crashes with an error like

warning: no more registered memory region table entries (max is 4096). Change using CHPL_RT_COMM_UGNI_MAX_MEM_REGIONS.
<internal>: error: Out of memory allocating "tasking layer unspecified data"

If I crank that environment variable up to 65536, the same computation seems to run fine. Is there a "safe" value that I can set this env variable to for all our users? Or is this indicative of a more serious problem that I'm papering over?

**Comment by ronawho** [comment]
tl;dr: If increasing it seems to work for you, there's no major downside for the types of comm patterns arkouda has. node size / hugepage size is the most memory regions you could ever need. So for instance on 256 GB nodes with 16M hugepages, 16384 is the most you could ever need.

The downside of increasing the value is that there is a numLocales*max_regions memory footprint per locale and there is an overhead for finding a memory region when doing comm, but arkouda mostly does large bulk comm so that overhead is negligible. 

---

The Aries NIC requires that memory is registered before you can do RDMA. ugni uses dynamic memory registration so that we can get good NUMA affinity. However, dynamic memory registration requires us to track which regions of memory are registered. There is one region for arrays larger than 2 hugepages, and for everything else 1 region per hugepage. 

CHPL_RT_COMM_UGNI_MAX_MEM_REGIONS is how many regions can be registered. Each node has to track every other nodes registered regions so there is a memory overhead for storing those tables and looking through the table when performing RDMA takes longer the larger the table is. For those reasons we tried to keep it on the smaller side. I think we sized it thinking that on 128 GB nodes with 16 MB hugepages, the most you could ever need is 8192, but we assumed that most programs would have several large arrays, which only require a single memory region each.

We (chapel folks) should probably size it based on node size and hugepage size automatically instead of hardcoding it to 4096.

**Comment by reuster986** [comment]
Thanks @ronawho , good to know! I will close this since it does not require a change to arkouda.

**Comment by ronawho** [comment]
Sure thing. Do you know what node size / hugepages are being used?

**Comment by reuster986** [comment]
256 GB nodes with 16 M hugepages. I think I'll just set it to 16384 memory regions.

**Comment by ronawho** [comment]
Sounds good. I'll open a Chapel issue for bumping it on our end. 

If you have a reproducer for this, would you mind sending it my way? I would expect arkouda to be using a fair number of large arrays, so I'm a little surprised that you ran out of descriptors with 16M hugepages. I wouldn't mind digging a little further just for my own peace of mind. (It could also indicate there's a lot of non-large-array allocations, which depending on the use-case means there might be an area to optimization allocation for.)

**Comment by reuster986** [comment]
Sorry, I just logged out, but I'm pretty sure I was running things like

```bash
test/UnitTestSubstringSearch -nl 120 --N=100_000_000
```

and it was throwing the error before it finished generated the random strings, I believe. I can get you a more precise reproducer on Monday, if needed.

**Comment by ronawho** [comment]
Just FYI we bumped the default value to 16K in https://github.com/chapel-lang/chapel/pull/15061, so after 1.21 is released (and you're using it) you shouldn't have to set `CHPL_RT_COMM_UGNI_MAX_MEM_REGIONS` anymore
